crates/sui-indexer/src/models/watermarks.rs

```diff
@@ -0,0 +1,217 @@
+        vec![
+            Watermark::upper_bound(WatermarkEntity::Checkpoints, epoch_hi, cp_hi),
+            Watermark::upper_bound(WatermarkEntity::Epochs, epoch_hi, epoch_hi),
+            Watermark::upper_bound(WatermarkEntity::Events, epoch_hi, tx_hi),
+            Watermark::upper_bound(WatermarkEntity::ObjectsHistory, epoch_hi, cp_hi),
+            Watermark::upper_bound(WatermarkEntity::Transactions, epoch_hi, tx_hi),
+        ]
+    }
+
+    pub fn new_lower_bounds(epoch_lo: u64, cp_lo: u64, tx_lo: u64) -> Vec<Watermark> {
+        vec![
+            Watermark::lower_bound(WatermarkEntity::Checkpoints, epoch_lo, cp_lo),
+            Watermark::lower_bound(WatermarkEntity::Epochs, epoch_lo, epoch_lo),
+            Watermark::lower_bound(WatermarkEntity::Events, epoch_lo, tx_lo),
+            Watermark::lower_bound(WatermarkEntity::ObjectsHistory, epoch_lo, cp_lo),
+            Watermark::lower_bound(WatermarkEntity::Transactions, epoch_lo, tx_lo),
+        ]
+    }
+
+    pub fn update_type(&self) -> WatermarkUpdateType {
+        match self.update {
+            WatermarkUpdate::UpperBound { .. } => WatermarkUpdateType::UpperBound,
+            WatermarkUpdate::LowerBound { .. } => WatermarkUpdateType::LowerBound,
+            WatermarkUpdate::LowerBoundTail { .. } => WatermarkUpdateType::LowerBoundTail,
+        }
+    }
+}
+
+impl From<Watermark> for StoredWatermark {
+    fn from(watermark: Watermark) -> Self {
+        match watermark.update {
+            WatermarkUpdate::UpperBound { epoch_hi, hi } => StoredWatermark {
+                entity: watermark.entity.as_str().to_string(),
+                epoch_hi: epoch_hi as i64,
+                hi: hi as i64,
+                ..StoredWatermark::default()
+            },
+            WatermarkUpdate::LowerBound { epoch_lo, lo } => StoredWatermark {
+                entity: watermark.entity.as_str().to_string(),
+                epoch_hi: epoch_lo as i64,
+                epoch_lo: epoch_lo as i64,
+                hi: lo as i64,
+                lo: lo as i64,
+                ..StoredWatermark::default()
+            },
+            WatermarkUpdate::LowerBoundTail { pruned_lo } => StoredWatermark {
+                entity: watermark.entity.as_str().to_string(),
+                pruned_lo: Some(pruned_lo as i64),
+                ..StoredWatermark::default()
+            },
+        }
+    }
+}
+
+impl From<StoredWatermark> for WatermarkRead {
+    fn from(watermark: StoredWatermark) -> Self {
+        WatermarkRead {
+            entity: match watermark.entity.as_str() {
+                "transactions" => WatermarkEntity::Transactions,
+                "objects_history" => WatermarkEntity::ObjectsHistory,
+                "checkpoints" => WatermarkEntity::Checkpoints,
+                "epochs" => WatermarkEntity::Epochs,
+                "events" => WatermarkEntity::Events,
+                _ => unreachable!(),
```

Comment:
This is a risk -- quite easy to add an entity and forget to update this (or run an old version of the software and trip up on this). We need a more robust way to handle that case.